test/utils/TestUniversalResolver.js

```diff
@@ -0,0 +1,101 @@
+
+contract('UniversalResolver', function (accounts) {
+
+    let ens, publicResolver, universalResolver, dummyOffchainResolver, nameWrapper;
+
+    beforeEach(async () => {
+        node = namehash.hash('eth');
+        ens = await ENS.new();
+        nameWrapper = await NameWrapper.new();
+        publicResolver = await PublicResolver.new(ens.address, nameWrapper.address, ZERO_ADDRESS, ZERO_ADDRESS);
+        universalResolver = await UniversalResolver.new(ens.address);
+        dummyOffchainResolver = await DummyOffchainResolver.new();
+
+        await ens.setSubnodeOwner('0x0', sha3('eth'), accounts[0], {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('eth'), sha3('test'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('test.eth'), publicResolver.address, {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('test.eth'), sha3('sub'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('sub.test.eth'), accounts[1], {from: accounts[0]});
+        await publicResolver.methods["setAddr(bytes32,address)"](namehash.hash('test.eth'), accounts[1], {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('test.eth'), sha3('offchain'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('offchain.test.eth'), dummyOffchainResolver.address, {from: accounts[0]})
+    });
+
+    describe('findResolver()', () => {
+        it('should find an exact match resolver', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('test.eth'));
+            expect(result).to.equal(publicResolver.address);
+        });
+
+        it('should find a resolver on a parent name', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('foo.test.eth'));
+            expect(result).to.equal(publicResolver.address);
+        });
+
+        it('should choose the resolver closest to the leaf', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('sub.test.eth'));
+            expect(result).to.equal(accounts[1]);
+        });
+    });
+
+    describe('resolve()', () => {
+        it('should resolve a record if `supportsInterface` throws', async () => {
+            const legacyResolver = await LegacyResolver.new();
+            await ens.setSubnodeOwner(namehash.hash('eth'), sha3('test2'), accounts[0], {from: accounts[0]});
+            await ens.setResolver(namehash.hash('test2.eth'), legacyResolver.address, {from: accounts[0]});
+            const data = (await legacyResolver.methods['addr(bytes32)'].request(namehash.hash('test.eth'))).data;
+            const result = await universalResolver.resolve(dns.hexEncodeName('test2.eth'), data);
+            const [ret] = ethers.utils.defaultAbiCoder.decode(["address"], result);
+            expect(ret).to.equal(legacyResolver.address);
```

Comment:
If I read the test correctly, `address` field returns the resolver address if the resolver doesn't support wildcard (='resolve' function), but returns the address record if it does support? How do you distinguish if it is returning the resolver address or eth address record?

Reply:
No, it always returns the result of calling `resolve` or `addr`. In this case the legacy resolver simply returns its own address.